BP_EXT/2022_001 - Google Groups are Enabled

taylorjstacey opened the pull request:
This pull request includes three changes.

1. Adds `has_authenticator_group_enabled` to the GKE query class. This checks to see if the `['authenticatorGroupsConfig']['enabled'] `key exists. If authentictor groups were previously enabled, the previously used gke-security-groups email will persist but the enabled key will disappear.
2. Added in the lint check for BP/2021_002. 
3. Provides the documentation and justification for why Google Groups for RBAC should be considered a best practice.

taylorjstacey commented:
@googlebot I signed it!

schweikert commented:
Thanks a lot Taylor for this pull request and welcome to the project!

schweikert commented on website/content/en/rules/gke/BP/2021_002.md:
```diff
@@ -0,0 +1,48 @@
+
+### Description
+
+IAM and Kubernetes RBAC work together to help manage access to your cluster.
```
We should whenever possible reference external documentation instead of adding it here (and the text seems to be copied from the public docs). For rule pages I would focus on the following:

- Why was the rule triggered
- Short description about why this is a recommended best-practice (i.e. the same description that you have in the rule implementation "Enable Google Groups for RBAC so cluster...".
- Hints about how to solve the issue (possibly also just pointing to public docs)
- Links to public documentation for further information.

schweikert replied:
This document still contains too much non-original content in my opinion. I would remove even more text.

taylorjstacey replied:
Reduced text.

schweikert commented on gcpdiag/queries/gke.py:
```diff
@@ -268,6 +268,9 @@ def has_logging_enabled(self) -> bool:
   def has_monitoring_enabled(self) -> bool:
     return self._resource_data['monitoringService'] != 'none'
 
+  def has_authenticator_group_enabled(self) -> bool:
```
Could you add a test for this new method in gke_test.py?

taylorjstacey replied:
I've written the test but it'll require additional data in the json dumps so I'll reach out to dwes@google.com as instructed in the README. Will update accordingly.

schweikert commented on gcpdiag/lint/gke/bp_2021_002_groups_enabled.py:
```diff
@@ -0,0 +1,37 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Google Groups for RBAC enabled.
```
I feel like this rule might be triggered too frequently and be possibly annoying for gcpdiag users (I know for instance that I personally rarely enable this). Is it really the case that we always recommend to use Google Groups for RBAC? I am not sure, especially also considering that even Autopilot doesn't have it enabled. I'll ask the opinion of some more folks on this.

taylorjstacey replied:
Adding in a temporary condition that this check won't apply to Autopilot clusters makes sense and can be added.

I don't think we should be too concerned about the rule being triggered too frequently because users can skip categories of lint rules as needed. 

I think it's worth adding the check for two reasons:
1. If we use gke/BP/2021_001 as an example, I know of scenarios why that wouldn't be enabled on GKE clusters just like Groups for RBAC. However, I would recommend both for production clusters far more often than not for reasons outlined in the documentation.
2. Groups for RBAC on GKE requires extra setup that isn't always required on other clouds, to enable group permission management specifically, so the check can be a good learning opportunity for new administrators coming into GCP.

schweikert replied:
gke/BP/2021_001 is not comparable in my opinion, because we definitely always recommend to enable Cloud Ops monitoring+logging, at least for system workloads. One guiding principle is that we are going to recommend everything that improves supportability of a product. Another example for this is recommending to enable serial output logging for GCE instances. There should be rarely a reason not to fulfill this best-practice.

Groups for RBAC is a complex feature that requires access at the organizational level, which I think most GKE administrators don't have (I am actually struggling right now to create a cluster where you could test this feature). I am still gathering feedback on it, but in my opinion this is going to be generating too many positives. In the end, best-practice rules are going to be opinionated by nature.

It is true that it is possible to exclude a class of rules or this one specifically, but every time that we implement a new rule it is always a tradeoff between the potential of raising too many issues that the user doesn't care about, and raising something that they might be interested in.

I wonder whether we should have a new class of rules that are disabled by default, like "HINT" or something like that. Rules that raise things that might be interesting for you, especially to learn about things. What do you think?

schweikert replied:
We do have that now: BP_EXT is for best-practice rules that are disabled by default.

taylorjstacey replied:
Updated to BP_EXT

schweikert replied:
Thanks Taylor. There is a code conflict that remains, but otherwise I think this is ready to be merged.

taylorjstacey replied:
Merge conflict should be resolved now.

schweikert commented on gcpdiag/lint/gke/bp_2021_002_groups_enabled.py:
```diff
@@ -0,0 +1,37 @@
+  for _, c in sorted(clusters.items()):
+    disabled = []
+    if not c.has_authenticator_group_enabled():
+      disabled.append('groups')
```
I don't understand why you are using this list here, what is the purpose of it?

taylorjstacey replied:
You're right, that had no purpose. Updated and committed.

SergeyKanzhelev commented:
this is CIS 6.8.3 https://cloud.google.com/kubernetes-engine/docs/concepts/cis-benchmarks so I would suggest we take this rule. The only question is how to categorize it so it wouldn't be too noisy.

@schweikert do you think we might need another category of rules? Maybe CIS as a category by itself?

taylorjstacey commented:
Added tests as needed and I greatly reduced to the text in the webpage summary for BP/2021_02 to make it more actionable.

Took @schweikert's note on the unnecessary list and removed that step.

schweikert commented on gcpdiag/lint/gke/snapshots/BP_EXT_2022_001.txt:
```diff
@@ -0,0 +1,3 @@
+*  gke/BP_EXT/2022_001: Google Groups for RBAC enabled.
+   (Error: 'authenticatorGroupsConfig')                                   [SKIP]
```
This looks weird. What does this error mean?

taylorjstacey replied:
Looks like there was an error with the unit test and it wasn't successfully catching a bug in the main check. This issue should be resolved now.
